src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -177,6 +177,8 @@ internal void SubscribeExternalErrorDiagnosticUpdateSourceToSolutionBuildEvents(
                     }
                     else
                     {
+                        // A real build just finished.  Clear out any results from the last "run code analysis" command.
+                        this.Services.GetRequiredService<ICodeAnalysisDiagnosticAnalyzerService>().Clear();
```

Comment:
> .Clear()

Probably this method should be called `OnBuildCompleted` to not leak implementation details of this service. I think it might also be better to move this code into `ExternalErrorDiagnosticUpdateSource.OnSolutionBuildCompleted()` instead of polluting `VisualStudioWorkspaceImpl`

Reply:
Actually - thinking a bit more, this operation should likely be triggered here: https://github.com/dotnet/roslyn/blob/918010f3d36ad243e91432c4b6d2b64a6bc94dfe/src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs#L171. Everything is already hooked up to a VS interface which has an API that lets us know when to clear prior errors, so we can just do it per-project with this API

Reply:
So, I believe the new method on `ICodeAnalysisDiagnosticAnalyzerService` should be `void Clear(projectId);` and should be triggered from `ExternalErrorDiagnosticUpdateSource.ClearErrors(projectId)`

Reply:
Thanks!  I'll look into this tomorrow!